File: 장세훈/LeetCode/medium/Permutations.java
class Permutations {
    /*
        https://leetcode.com/problems/permutations/description/
        46. Permutations

        주어진 숫자들로 만들 수 있는 모든 순열을 만드는 문제
        dfs로 완전탐색하며 모든 조합을 만든다.
        visit를 통해 현재 숫자의 사용 여부를 저장한다.
     */


    List<List<Integer>> answer = new ArrayList<>();

    public List<List<Integer>> permute(int[] nums) {
        boolean[] visit = new boolean[nums.length];
        dfs(nums, new ArrayList<>(), visit);
        return answer;
    }

    void dfs(int[] nums, List<Integer> list, boolean[] visit) {
        if (list.size() == nums.length) {
            answer.add(new ArrayList<>(list));
            return;
        }
        for (int i = 0; i < nums.length; i++) {
            if (!visit[i]) {
                visit[i] = true;
                list.add(nums[i]);
                dfs(nums, list, visit);
                list.removeLast();
                visit[i] = false;
            }
        }
    }
}

File: 정수빈/LeetCode/medium/permutations.py
# https://leetcode.com/problems/permutations/
# 설명: permute을 백트레킹으로 구현하는 문제입니다. 
# 난이도: medium
# 시간: 15분 30초

class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        
        ans = []
        def bk(case = [], nnums = []):
            if not nnums:
                ans.append(case)
                return 

            for i in range(len(nnums)):
                new_nums = nnums[:]
                n = nnums[i]
                new_nums.remove(n)
                bk(i+1, case + [nnums[i]], new_nums)
            
        bk(0, [], nums)
        return ans